feat: add `blas/base/wasm/scasum`

ShabiShett07 opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: passed
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

Progresses: 2039

## Description

> What is the purpose of this pull request?

This pull request:

-   adds package `blas/base/wasm/scasum`
## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #2039 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

ShabiShett07 committed:
feat: add blas/base/wasm/scasum

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

ShabiShett07 commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

ShabiShett07 committed:
fix: markdown example

Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 committed:
fix: package.json lint errors

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/README.md:
````diff
@@ -0,0 +1,304 @@
+var scasum = require( '@stdlib/blas/base/wasm/scasum' );
+```
+
+#### scasum.main( N, cx, stride )
````
@ShabiShett07 Given the variable name changes you made in the `blas/base/*`, would you mind updating from `cx` to `x` throughout this PR?

kgryte replied:
```suggestion
#### scasum.main( N, cx, strideX )
```

Let's also make this `strideX` to match the docs below.

ShabiShett07 committed:
chore: change variable naming

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/README.md:
```diff
@@ -0,0 +1,304 @@
+
+-   **offsetX**: starting index for `x`.
+
+While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameter supports indexing semantics based on a starting index. For example, to compute the sum of last three elements,
```
```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameter supports indexing semantics based on a starting index. For example, to compute the sum of the last three elements,
```

kgryte committed:
docs: fix missing word

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/README.md:
```diff
@@ -0,0 +1,304 @@
+
+var sum = scasum.ndarray( 2, x, 1, x.length-2 );
+// returns 8.0
+
+// Using a negative stride to sum from the last element:
+sum = scasum.ndarray( 4, x, -1, x.length-1 );
+// returns 19.0
```
```suggestion
```

kgryte committed:
docs: update example

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/README.md:
```diff
@@ -0,0 +1,300 @@
+
+var x = new Complex64Array( [ -2.0, 1.0, 3.0, -5.0, 4.0, 0.0, -1.0, -3.0 ] );
+
+var sum = scasum.ndarray( 2, x, 1, x.length-2 );
+// returns 8.0
```
```suggestion
var sum = scasum.ndarray( 3, x, 1, x.length-2 );
// returns 16.0
```

kgryte committed:
docs: update example to match docs

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/README.md:
```diff
@@ -0,0 +1,300 @@
+
+#### scasum.Module.prototype.main( N, xp, sx )
+
+Computes the sum of absolute values.
```
This description is incomplete. See above.

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/README.md:
```diff
@@ -0,0 +1,300 @@
+
+#### scasum.Module.prototype.ndarray( N, xp, sx, ox )
+
+Computes the sum of absolute values using alternative indexing semantics.
```
This description is incomplete. See above.

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/benchmark/benchmark.module.main.js:
```diff
@@ -0,0 +1,137 @@
+
+		// Reallocate the underlying memory to allow storing one vector:
+		nb = bytesPerElement( options.dtype );
+		mod.realloc( (N*nb)+nb );
```
```suggestion
		mod.realloc( N*nb );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/benchmark/benchmark.module.ndarray.js:
```diff
@@ -0,0 +1,137 @@
+
+		// Reallocate the underlying memory to allow storing one vector:
+		nb = bytesPerElement( options.dtype );
+		mod.realloc( (N*nb)+nb );
```
```suggestion
		mod.realloc( N*nb );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/benchmark/benchmark.module.main.js:
```diff
@@ -0,0 +1,137 @@
+		// Define a pointer (i.e., byte offset) for storing the input vector:
+		xptr = 0;
+
+		x = new Complex64Array( uniform( N, -10.0, 10.0, options ) );
```
```suggestion
		// Generate an array of random complex numbers:
		x = new Complex64Array( uniform( N, -10.0, 10.0, options ) );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/benchmark/benchmark.module.ndarray.js:
```diff
@@ -0,0 +1,137 @@
+		// Define a pointer (i.e., byte offset) for storing the input vector:
+		xptr = 0;
+
+		x = new Complex64Array( uniform( N, -10.0, 10.0, options ) );
```
```suggestion
		// Generate an array of random complex numbers:
		x = new Complex64Array( uniform( N, -10.0, 10.0, options ) );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/docs/repl.txt:
```diff
@@ -0,0 +1,499 @@
+    > s = {{alias}}.main( 2, x, 2 )
+    7.0
+
+    // Use view offset; e.g., starting at 2nd element:
```
```suggestion
    // Using typed array views:
```

Don't go rogue. Be consistent with other packages.

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/docs/repl.txt:
```diff
@@ -0,0 +1,499 @@
+    7.0
+
+    // Use view offset; e.g., starting at 2nd element:
+    > var x0 = new {{alias:@stdlib/array/complex64}}([1.0,-2.0,3.0,-4.0,5.0,-6.0,7.0,-8.0 ]);
```
```suggestion
    > var x0 = new {{alias:@stdlib/array/complex64}}( [ 1.0, -2.0, 3.0, -4.0, 5.0, -6.0 ] );
```

ShabiShett07 replied:
This was done intentionally thinking that `max-len` problem was caused by this, but now when rechecking it, it is causing the problem of `max-len`, so left it as it is

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/docs/repl.txt:
```diff
@@ -0,0 +1,499 @@
+    > var x0 = new {{alias:@stdlib/array/complex64}}([1.0,-2.0,3.0,-4.0,5.0,-6.0,7.0,-8.0 ]);
+    > var x1 = new {{alias:@stdlib/array/complex64}}( x0.buffer, x0.BYTES_PER_ELEMENT*1 );
+    > s = {{alias}}.main( 2, x1, 2 )
+    22.0
```
```suggestion
    > s = {{alias}}.main( 2, x1, 1 )
    18.0
```

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/docs/repl.txt:
```diff
@@ -0,0 +1,499 @@
+
+{{alias}}.ndarray( N, x, strideX, offsetX )
+    Computes the sum of the absolute values of the real and imaginary
+    components of a single-precision complex floating-point vector
+    using alternative indexing semantics.
```
Text wrapping is off.

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/docs/repl.txt:
```diff
@@ -0,0 +1,499 @@
+    > var s = {{alias}}.ndarray( x.length, x, 1, 0 )
+    15.0
+
+    // Using offset parameter:
```
```suggestion
    // Advanced indexing:
```

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/docs/repl.txt:
```diff
@@ -0,0 +1,499 @@
+
+
+{{alias}}.Module.prototype.ndarray( N, xp, sx, ox )
+    Computes the sum of the absolute values of the real and imaginary
```
Your text wrapping is off here. Everything should be wrapped to 80 chars. Meaning `components` should be moved up and everything rewrapped. Please check the other descriptions in this file.

ShabiShett07 replied:
I think moving `components` will cause the problem as the line length will be `81`

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/docs/types/index.d.ts:
```diff
@@ -0,0 +1,317 @@
+	* // Specify a vector length:
+	* var N = 4;
+	*
+	* // Define pointer (i.e., byte offsets) for storing two vectors:
```
```suggestion
	* // Define a pointer (i.e., byte offset) for storing one vector:
```

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/docs/types/index.d.ts:
```diff
@@ -0,0 +1,317 @@
+	* // Specify a vector length:
+	* var N = 4;
+	*
+	* // Define pointer (i.e., byte offsets) for storing two vectors:
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/docs/types/index.d.ts:
```diff
@@ -0,0 +1,317 @@
+	* // Specify a vector length:
+	* var N = 4;
+	*
+	* // Define pointer (i.e., byte offsets) for storing the input vector:
```
```suggestion
	* // Define a pointer (i.e., byte offset) for storing one input vector:
```

This and similar changes throughout this PR.

kgryte reviewed (changes requested):
Left an initial comments after reviewing a few files. I suggest going back through and doing a line-by-line check to ensure that everything is cleaned up.

ShabiShett07 committed:
chore: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: minor clean-up

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 committed:
chore: minor clean-up

Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 committed:
chore: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 commented:
@kgryte, made the necessary changes you have mentioned, as well as there were other changes needed too, but one query is still remaining that is mentioned above.

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/docs/repl.txt:
```diff
@@ -0,0 +1,499 @@
+    7.0
+
+    // Using typed array views:
+    > var x0 = new {{alias:@stdlib/array/complex64}}([1.0,-2.0,3.0,-4.0,5.0,-6.0,7.0,-8.0 ]);
```
```suggestion
    > var x0 = new {{alias:@stdlib/array/complex64}}( [1.0,2.0,3.0,-4.0,5.0,-6.0,7.0,8.0] );
```

kgryte committed:
style: adjust spacing

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/docs/types/index.d.ts:
```diff
@@ -0,0 +1,317 @@
+	* // Specify a vector length:
+	* var N = 4;
+	*
+	* // Define a pointer (i.e., byte offset) for storing one vectors:
```
```suggestion
	* // Define a pointer (i.e., byte offset) for storing one vector:
```

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/docs/types/index.d.ts:
```diff
@@ -0,0 +1,317 @@
+	* // Specify a vector length:
+	* var N = 4;
+	*
+	* // Define a pointer (i.e., byte offset) for storing one vectors:
```
```suggestion
	* // Define a pointer (i.e., byte offset) for storing one vector:
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/docs/types/index.d.ts:
```diff
@@ -0,0 +1,317 @@
+	* // Specify a vector length:
+	* var N = 4;
+	*
+	* // Define a pointer (i.e., byte offset) for storing one vector:
```
```suggestion
	* // Define a pointer (i.e., byte offset) for storing the input vector:
```

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/docs/types/index.d.ts:
```diff
@@ -0,0 +1,317 @@
+	* // Specify a vector length:
+	* var N = 4;
+	*
+	* // Define a pointer (i.e., byte offset) for storing one vector:
```
```suggestion
	* // Define a pointer (i.e., byte offset) for storing the input vector:
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/test/test.main.js:
```diff
@@ -0,0 +1,162 @@
+	var N;
+
+	x = new Complex64Array([
+		-2.0,  // 0
```
```suggestion
		-2.0, // 0
```

@ShabiShett07 Throughout this file, would you mind aligning these comments?

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/test/test.main.js:
```diff
@@ -0,0 +1,162 @@
+	var N;
+
+	x = new Complex64Array([
+		-2.0,  // 1
```
```suggestion
		-2.0, // 1
```

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/test/test.module.main.js:
```diff
@@ -0,0 +1,162 @@
+	xp = 0;
+
+	mod.write( xp, new Complex64Array([
+		-2.0,  // 0
```
```suggestion
		-2.0, // 0
```

Same comment. Always keep these comments aligned.

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/test/test.module.ndarray.js:
```diff
@@ -0,0 +1,230 @@
+	xp = 0;
+
+	mod.write( xp, new Complex64Array([
+		-2.0,  // 0
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/test/test.ndarray.js:
```diff
@@ -0,0 +1,157 @@
+	var N;
+
+	x = new Complex64Array([
+		-2.0,  // 0
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/test/test.ndarray.js:
```diff
@@ -0,0 +1,157 @@
+	t.end();
+});
+
+tape( 'the `main` method computes the sum of the absolute values of the real and imaginary components of a single-precision complex floating-point vector', function test( t ) {
```
```suggestion
tape( 'the `ndarray` method computes the sum of the absolute values of the real and imaginary components of a single-precision complex floating-point vector', function test( t ) {
```

This change applies throughout this file.

kgryte reviewed (changes requested):
Other than some tidiness issues, this PR is looking good.

ShabiShett07 committed:
chore: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 commented:
@kgryte, the test files required some changes as you said and I have considered and finished out the clean up

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/scasum/test/test.main.js:
```diff
@@ -0,0 +1,162 @@
+	t.end();
+});
+
+tape( 'the `main` method supports view offset', function test( t ) {
```
```suggestion
tape( 'the `main` method supports view offsets', function test( t ) {
```

kgryte committed:
test: fix description

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
